src/framework/rendering/Model3D.cpp

```diff
@@ -0,0 +1,268 @@
+   glEnableVertexAttribArray(vertexAttributeColor);
+   glVertexAttribPointer(vertexAttributeColor, 3, GL_FLOAT, GL_FALSE, 0, 0);
+
+   glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+   return true;
+}
+
+//---------------------------------------------------------------------------------------------------
+bool Model3D::CreateTextureBuffer(vector<vec3>& data)
+{
+   if(data.size() <= 0)
+      return false;
+
+   if(m_BufferIds[vertexAttributeTexture] != 0) {
+      glBindBuffer(GL_ARRAY_BUFFER, 0);
+      glDeleteBuffers(1, &m_BufferIds[vertexAttributeTexture]);
+   }
+
+   glGenBuffers(1, &m_BufferIds[vertexAttributeTexture]);
+   glBindBuffer(GL_ARRAY_BUFFER, m_BufferIds[vertexAttributeTexture]);
+   glBufferData(GL_ARRAY_BUFFER, data.size() * sizeof(vec3), data.data(), GL_STATIC_DRAW);
+   glEnableVertexAttribArray(vertexAttributeTexture);
+   glVertexAttribPointer(vertexAttributeTexture, 3, GL_FLOAT, GL_FALSE, 0, 0);
+
+   glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+   return true;
+}
+
+//---------------------------------------------------------------------------------------------------
+bool Model3D::CreateNormalBuffer(vector<vec3>& data)
+{
+   if(data.size() <= 0)
+      return false;
+
+   if(m_BufferIds[vertexAttributeNormal] != 0) {
+      glBindBuffer(GL_ARRAY_BUFFER, 0);
+      glDeleteBuffers(1, &m_BufferIds[vertexAttributeNormal]);
+   }
+
+   glGenBuffers(1, &m_BufferIds[vertexAttributeNormal]);
+   glBindBuffer(GL_ARRAY_BUFFER, m_BufferIds[vertexAttributeNormal]);
+   glBufferData(GL_ARRAY_BUFFER, data.size() * sizeof(vec3), data.data(), GL_STATIC_DRAW);
+   glEnableVertexAttribArray(vertexAttributeNormal);
+   glVertexAttribPointer(vertexAttributeNormal, 3, GL_FLOAT, GL_FALSE, 0, 0);
+
+   glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+   return true;
+}
+
+//---------------------------------------------------------------------------------------------------
+bool Model3D::CreateTangentBuffer(vector<vec4>& data)
+{
+   if(data.size() <= 0)
+      return false;
+
+   if(m_BufferIds[vertexAttributeTangent] != 0) {
+      glBindBuffer(GL_ARRAY_BUFFER, 0);
+      glDeleteBuffers(1, &m_BufferIds[vertexAttributeTangent]);
+   }
+
+   glGenBuffers(1, &m_BufferIds[vertexAttributeTangent]);
+   glBindBuffer(GL_ARRAY_BUFFER, m_BufferIds[vertexAttributeTangent]);
+   glBufferData(GL_ARRAY_BUFFER, data.size() * sizeof(vec4), data.data(), GL_STATIC_DRAW);
+   glEnableVertexAttribArray(vertexAttributeTangent);
+   glVertexAttribPointer(vertexAttributeTangent, 4, GL_FLOAT, GL_FALSE, 0, 0);
+
+   glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+   return true;
+}
+
+//---------------------------------------------------------------------------------------------------
+bool Model3D::CreateVAO()
+{
+   if(m_vaoId != 0) {
+      glBindVertexArray(0);
+      glDeleteVertexArrays(1, &m_vaoId);
+   }
+
+   glGenVertexArrays(1, &m_vaoId);
+   glBindVertexArray(m_vaoId);
+
+   if(m_BufferIds[vertexAttributePosition] != 0) {

```

Comment:
you may want to use these in loop. However its just a suggestion. (The current approach is more readable though)